File: packages/node/src/daemon/daemon-main.ts
#!/usr/bin/env node

import http from 'node:http';
import { BridgeService } from '../core/bridge-service.js';
import { attachDaemonApi } from './http-api.js';
import { getConfiguredPort } from '../utils/port.js';
import { logT, portLang } from '../i18n/lang.js';
import { logger, setLogLevel } from '../utils/logger.js';
import { PACKAGE_VERSION } from '../version.js';

export async function runDaemon(): Promise<void> {
  const L = portLang();
  const service = new BridgeService({
    onAfterBridgeDrain: () => process.exit(0),
  });

  const httpServer = http.createServer();
  attachDaemonApi(service, httpServer);

  try {
    await service.startSharedHttpServer(httpServer);
  } catch (err) {
    const e = err as NodeJS.ErrnoException;
    const port = getConfiguredPort();
    if (e.code === 'EADDRINUSE') {
      logger.error(logT(L, 'daemonPortInUse', { port }));
    } else {
      logger.error(logT(L, 'failedToStartServer'), err);
    }
    await service.stop();
    process.exit(1);
    return;
  }

  const port = service.listeningPort;
  logger.info(
    logT(L, 'daemonListening', {
      url: `http://127.0.0.1:${port}`,
      version: PACKAGE_VERSION,
    }),
  );

  let shuttingDown = false;
  const shutdown = async () => {
    if (shuttingDown) return;
    shuttingDown = true;
    logger.info(logT(L, 'shuttingDownMcp'));
    await new Promise<void>((resolve, reject) => {
      httpServer.close((err) => (err ? reject(err) : resolve()));
    });
    await service.stop();
    process.exit(0);
  };

  process.on('SIGINT', () => void shutdown());
  process.on('SIGTERM', () => void shutdown());
}

/** Called from CLI router with argv already trimmed after `daemon`. */
export async function runDaemonFromCli(daemonArgv: string[]): Promise<void> {
  if (daemonArgv.includes('--debug')) {
    setLogLevel('debug');
  }
  await runDaemon();
}

File: packages/python/src/lionscraper/daemon/daemon_main.py
from __future__ import annotations

import asyncio
import signal
import sys

from lionscraper.core.bridge_service import BridgeService
from lionscraper.i18n.lang import log_t, port_lang
from lionscraper.utils.logger import logger, set_log_level
from lionscraper.utils.port import get_configured_port
from lionscraper.version import PACKAGE_VERSION


async def _run_daemon() -> None:
    stop = asyncio.Event()

    def after_drain() -> None:
        stop.set()

    service = BridgeService(on_after_bridge_drain=after_drain)

    try:
        await service.start()
    except OSError as e:
        port = get_configured_port()
        L = port_lang()
        errn = getattr(e, "winerror", None) or getattr(e, "errno", None)
        if errn in (98, 48, 10048) or "Address already in use" in str(e):
            logger.error(log_t(L, "daemonPortInUse", {"port": port}))
        else:
            logger.error(log_t(L, "failedToStartServer"), e)
        await service.stop()
        raise SystemExit(1) from e

    port = service.listening_port
    logger.info(
        log_t(
            port_lang(),
            "daemonListening",
            {"url": f"http://127.0.0.1:{port}", "version": PACKAGE_VERSION},
        )
    )

    async def shutdown_graceful() -> None:
        await service.stop()
        stop.set()

    def on_sig() -> None:
        asyncio.create_task(shutdown_graceful())

    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        if hasattr(signal, sig.name):
            try:
                loop.add_signal_handler(sig, on_sig)
            except NotImplementedError:
                pass

    await stop.wait()
    raise SystemExit(0)


def run_daemon_from_cli(daemon_argv: list[str]) -> None:
    if "--debug" in daemon_argv:
        set_log_level("debug")
    try:
        asyncio.run(_run_daemon())
    except SystemExit as e:
        raise e
